File: Seccion 16 - Arboles en C++/07_eliminar_nodo.cpp
// Insertar nuevos nodos al arbol

#include <iostream>
#include <stdlib.h>

using namespace std;

struct Nodo {
    int dato;
    Nodo *der;
    Nodo *izq;
    Nodo *padre;
};

// Prototipos de funcion
void menu();
Nodo *crearNodo(int, Nodo *);
void insertarNodo(Nodo *&, int, Nodo *);
void mostrarArbol(Nodo *, int);
bool buscarNodo(Nodo *, int);
void preOrden(Nodo *);
void inOrden(Nodo *);
void postOrden(Nodo *);
void eliminarNodo(Nodo *, int);
void eliminar(Nodo *);
Nodo *minimo(Nodo *);
void reemplazar(Nodo *, Nodo *);
void destruirNodo(Nodo *);

Nodo *arbol = NULL;

int main() {

    menu();

    system("pause");
    return 0;
}

// Funcion del menu
void menu() {
    int dato, opcion, contador = 0;

    do {
        cout << "\t.:MENU:." << endl;
        cout << "1. Insertar un nuevo nodo." << endl;
        cout << "2. Mostrar el árbol completo." << endl;
        cout << "3. Buscar un elemento en el árbol." << endl;
        cout << "4. Recorrer el árbol en PreOrden." << endl;
        cout << "5. Recorrer el árbol en InOrden." << endl;
        cout << "6. Recorrer el árbol en PostOrden." << endl;
        cout << "7. Eliminar un nodo del árbol." << endl;
        cout << "8. Salir." << endl;
        cout << "Opcion: ";
        cin >> opcion;

        switch (opcion) {
            case 1:
                cout << "\nDigite un numero: ";
                cin >> dato;
                insertarNodo(arbol, dato, NULL); // Insertamos un nuevo nodo
                cout << "\n";
                break;
            case 2:
                cout << "\nMostrando el arbol completo:\n\n";
                mostrarArbol(arbol, contador);
                cout << "\n";
                break;
            case 3:
                cout << "\nDigite el elemento a buscar: ";
                cin >> dato;
                if (buscarNodo(arbol, dato) == true) {
                    cout << "\nEl elemento " << dato << " ha sido encontrado en el arbol." << endl;
                } else {
                    cout << "\nElemento no encontrado.\n" << endl;
                }
                cout << "\n";
                break;
            case 4:
                cout << "\nRecorrido en PreOrden:\n";
                preOrden(arbol);
                cout << "\n\n";
                break;
            case 5:
                cout << "\nReccorrido en InOrden:\n";
                inOrden(arbol);
                cout << "\n\n";
                break;
            case 6:
                cout << "\nRecorrido en PostOrden:\n";
                postOrden(arbol);
                cout << "\n\n";
                break;
            case 7:
                cout << "\nDigite el número a eliminar: ";
                cin >> dato;
                eliminarNodo(arbol, dato);
                cout << "\n";
                break;
            
            default:
                break;
        }

        system("cls");
    } while (opcion != 8);
    
}

// Funcion para crear un Nodo
Nodo *crearNodo(int n, Nodo *padre) {
    Nodo *nuevo_nodo = new Nodo();

    nuevo_nodo->dato = n;
    nuevo_nodo->der = NULL;
    nuevo_nodo->izq = NULL;
    nuevo_nodo->padre = padre;

    return nuevo_nodo;
}

// Funcion para insertar elementos en el arbol
void insertarNodo(Nodo *&arbol, int n, Nodo *padre) {
    if (arbol == NULL) { // Si el arbol esta vacio 
        Nodo *nuevo_nodo = crearNodo(n, padre);
        arbol = nuevo_nodo;
    } else { // Si el arbol tiene un nodo o mas
        int valorRaiz = arbol->dato; // Obtenemos el valor de la raiz

        if (n < valorRaiz) {
            insertarNodo(arbol->izq, n, arbol); // Si el elemento es menor que la raiz
        } else {
            insertarNodo(arbol->der, n, arbol); // Si el elemento es mayor que la raiz
        }
    }
}

// Funcion para mostrar el arbol completo
void mostrarArbol(Nodo *arbol, int cont) {
    if (arbol == NULL) { // Si el arbol esta vacio
        return;
    } else { // Si el arbol no esta vacio
        mostrarArbol(arbol->der, cont+1);

        for (int i = 0; i < cont; i++) {
            cout << "   ";
        }
        cout << arbol->dato << endl;
        mostrarArbol(arbol->izq, cont+1);
    }
}

// Funcion para buscar un elemento en el arbol
bool buscarNodo(Nodo *arbol, int n) {
    if (arbol == NULL) { // si el arbol esta vacio
        return false;
    } else if (arbol->dato == n) {
        return true;
    } else if (n < arbol->dato) {
        return buscarNodo(arbol->izq, n);
    } else {
        return buscarNodo(arbol->der, n);
    }
}

// Función para recorrido en profundidad - PreOrden
void preOrden(Nodo *arbol) {
    if (arbol == NULL) { // Si el árbol está vacío
        return;
    } else {
        cout << arbol->dato << " - ";
        preOrden(arbol->izq);
        preOrden(arbol->der);
    }
}

// Función para recorrido en profundidad - InOrden
void inOrden(Nodo *arbol) {
    if (arbol == NULL) {
        return;
    } else {
        inOrden(arbol->izq);
        cout << arbol->dato << " - ";
        inOrden(arbol->der);
    }
}

// Función para recorrido en profundidad - PostOrden
void postOrden(Nodo *arbol) {
    if (arbol == NULL) {
        return;
    } else {
        postOrden(arbol->izq);
        postOrden(arbol->der);
        cout << arbol->dato << " - ";
    }
}

// Función para buscar un nodo y eliminarlo
void eliminarNodo(Nodo *arbol, int n) {
    if (arbol == NULL) { // Si el árbol está vacío
        return;
    } else if (n < arbol->dato) { // Si el valor es menor
        eliminarNodo(arbol->izq, n); // Busca por la izquierda
    } else if (n > arbol->dato) { // Si el valor es mayor
        eliminarNodo(arbol->der, n); // Busca por la derecha    
    } else { // Si ya encontraste el valor
        eliminar(arbol);
    }
}

// Función para determinar el nodo más izquierdo posible
Nodo *minimo(Nodo *arbol) {
    if (arbol == NULL) { // Si el árbol está vacío
        return NULL;
    }
    if (arbol->izq) { // Si tiene hijo izquierdo
        return minimo(arbol->izq); // Buscamos la parte más izquierda posible
    } else { // Si no tiene hijo zquierdo
        return arbol; // Retornamos el mismo nodo
    }
}

// Función para reemplazar un nodo por otro
void reemplazar(Nodo *arbol, Nodo *nuevo_nodo) {
    if (arbol->padre) {
        // arbol->padre hay que asignarle su nuevo hijo
        if (arbol->dato == arbol->padre->izq->dato) {
            arbol->padre->izq = nuevo_nodo;
        } else if (arbol->dato == arbol->padre->der->dato) {
            arbol->padre->der = nuevo_nodo;
        }
    }
    if (nuevo_nodo) {
        // Procedemos a asignarle su nuevo padre
        nuevo_nodo->padre = arbol->padre;
    }
}

// Función para destruir un nodo
void destruirNodo(Nodo *nodo) {
    nodo->izq = NULL;
    nodo->der = NULL;
    delete nodo;
}

// Función para eliminar el nodo encontrado
void eliminar(Nodo *nodoEliminar) {
    if (nodoEliminar->izq && nodoEliminar->der) {// Si el noto tiene hijos izq y der
        Nodo *menor = minimo(nodoEliminar->der);
        nodoEliminar->dato = menor->dato;
        eliminar(menor);
    } else if (nodoEliminar->izq) { // Si tiene hijo izquierdo
        reemplazar(nodoEliminar, nodoEliminar->izq);
        destruirNodo(nodoEliminar);
    } else if (nodoEliminar->der) { // Si tiene hijo derecho
        reemplazar(nodoEliminar, nodoEliminar->der);
        destruirNodo(nodoEliminar);
    } else { // No tiene hijos
        reemplazar(nodoEliminar, NULL);
        destruirNodo(nodoEliminar);
    }
}

File: Seccion 16 - Arboles en C++/Seccion_16_Python/07_eliminar_nodo.py

class Nodo:
    
    def __init__(self, dato=None, izq=None, der=None):
        self.dato = dato
        self.izq = izq
        self.der = der
        self.padre = None
    
    def __str__(self):
        return "{} -> ".format(self.dato)

class Arbol:
    
    def __init__(self):
        self.raiz = None
    
    def agregar_nodo(self, nodo: Nodo):
        if self.raiz == None:
            self.raiz = nodo
        else:
            aux = self.raiz
            padre = None
            
            while aux != None:
                padre = aux
                if nodo.dato >= aux.dato:
                    aux = aux.der
                else:
                    aux = aux.izq
            
            nodo.padre = padre
            if nodo.dato >= padre.dato:
                padre.der = nodo
            else:
                padre.izq = nodo
    
    def mostrar_arbol_completo(self, nodo: Nodo, cont: int):
        if nodo == None:
            print("", end="")
        else:
            self.mostrar_arbol_completo(nodo.der, cont+1)
            
            for i in range(cont):
                print("   ", end="")

            print(nodo.dato)
            self.mostrar_arbol_completo(nodo.izq, cont+1)
    
    def buscar_nodo(self, nodo: Nodo, n: int):
        if nodo == None:
            return False
        elif nodo.dato == n:
            return True
        elif n < nodo.dato:
            return self.buscar_nodo(nodo.izq, n)
        else:
            return self.buscar_nodo(nodo.der, n)
    
    def pre_orden(self, nodo: Nodo):
        if nodo == None:
            return ""
        else:
            print(nodo, end="")
            self.pre_orden(nodo.izq)
            self.pre_orden(nodo.der)
    
    def in_orden(self, nodo: Nodo):
        if nodo == None:
            return ""
        else:
            self.pre_orden(nodo.izq)
            print(nodo, end="")
            self.pre_orden(nodo.der)
    
    def post_orden(self, nodo: Nodo):
        if nodo == None:
            return ""
        else:
            self.pre_orden(nodo.izq)
            self.pre_orden(nodo.der)
            print(nodo, end="")
    
    def _minimo(self, nodo: Nodo):
        current = nodo
        while current.izq != None:
            current = current.izq
        
        return current

    def eliminar_nodo(self, nodo: Nodo, n: int):
        if self.raiz == None:
            print("No hay elementos en el árbol.")
        else:
            if n < nodo.dato:
                self.eliminar_nodo(nodo.izq, n)
            elif n > nodo.dato:
                self.eliminar_nodo(nodo.der, n)
            else:
                if nodo.izq != None and nodo.der != None:
                    menor = self._minimo(nodo.der)
                    nodo.dato = menor.dato
                    self.eliminar_nodo(nodo.der, menor.dato)
                elif nodo.der != None:
                    menor = self._minimo(nodo.der)
                    nodo.dato = menor.dato
                    self.eliminar_nodo(nodo.der, menor.dato)
                elif nodo.izq != None:
                    nodo.dato = nodo.izq.dato
                    self.eliminar_nodo(nodo.izq, nodo.izq.dato)
                else:
                    if nodo.padre.izq != None:
                        if nodo.padre.izq.dato == nodo.dato:
                            nodo.padre.izq = None
                    if nodo.padre.der != None:
                        if nodo.padre.der.dato == nodo.dato:
                            nodo.padre.der = None

    def get_raiz(self):
        return self.raiz

def menu():
    n: int
    opcion: int = 0
    contador: int = 0
    
    arbol = Arbol()
    
    while opcion != 8:
        print("\n.:MENU:.")
        print("1. Insertar un nuevo nodo.")
        print("2. Mostrar arbol completo.")
        print("3. Buscar un elemento en el árbol.")
        print("4. Recorrer el árbol en PreOrden.")
        print("5. Recorrer el árbol en InOrden.")
        print("6. Recorrer el árbol en PostOrden.")
        print("7. Eliminar un nodo del árbol.")
        print("8. Salir.")
        opcion = int(input("Opción: "))
        
        if opcion == 1:
            n = int(input("\nDigite un número entero: "))
            nodo = Nodo(dato=n)
            arbol.agregar_nodo(nodo)
        elif opcion == 2:
            print("\nMostrando el árbol completo:")
            arbol.mostrar_arbol_completo(arbol.get_raiz(), contador)
        elif opcion == 3:
            elemento = int(input("\nDigite el elemento a buscar: "))
            if arbol.buscar_nodo(arbol.get_raiz(), elemento):
                print("El elemento {} SÍ se encuentra en el árbol.".format(elemento))
            else:
                print("El elemento {} NO se encuentra en el árbol.".format(elemento))
        elif opcion == 4:
            print("\nRecorrido del árbol en PreOrden:")
            arbol.pre_orden(arbol.get_raiz())
        elif opcion == 5:
            print("\nRecorrido del árbol en InOrden:")
            arbol.in_orden(arbol.get_raiz())
        elif opcion == 6:
            print("\nRecorrido del árbol en PostOrden:")
            arbol.post_orden(arbol.get_raiz())
        elif opcion == 7:
            dato = int(input("\nDigite el elemento que desea eliminar: "))
            arbol.eliminar_nodo(arbol.get_raiz(), dato)
            print()

def main():
    menu()

if __name__ == "__main__":
    main()